### PR TITLE
Tweak install attributes to differentiate between project and client

### PR DESCRIPTION
**Pull request by asteflova**
First attempt at tweaking the install attributes. The goal is to be able to differentiate between commands run on the server (like {project-package-install}) and the client (like {client-package-install}).

* [ ] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [ ] Foreman 3.7/Katello 4.9 (planned Satellite 6.14)
* [ ] Foreman 3.6/Katello 4.8
* [ ] Foreman 3.5/Katello 4.7 (Satellite 6.13)
* [ ] Foreman 3.4/Katello 4.6 (EL8 only)
* [ ] Foreman 3.3/Katello 4.5 on EL7 & EL8 (Satellite 6.12 on EL8 only; orcharhino 6.4 on EL8 only)
* [ ] Foreman 3.2/Katello 4.4 on EL7 & EL8
* [ ] Foreman 3.1/Katello 4.3 on EL7 & EL8 (Satellite 6.11 EL7/8; orcharhino 6.3 on EL7/8)
* For Foreman 3.0 or older, please create a separate PR.
* We do not accept PRs for Foreman 2.4 or older.

### CODE REVIEW

File: guides/common/modules/con_troubleshooting-postgresql.adoc [hunk 1]
```diff
@@ -30,7 +30,7 @@ If a database was previously created using PostgreSQL 10, perform an upgrade:
 +
 [options="nowrap" subs="+quotes,attributes"]
 ----
-# {package-install} postgresql-upgrade
+# {client-package-install} postgresql-upgrade
```
Comment (ekohl) [inline guides/common/modules/con_troubleshooting-postgresql.adoc:33]: This is on the server, so I think `project-package-install` is more appropriate:
```suggestion
# {project-package-install} postgresql-upgrade
```

File: guides/common/modules/proc_adding-rhel-system-roles.adoc [hunk 2]
```diff
@@ -33,7 +33,7 @@ endif::[]
 +
 [options="nowrap" subs="+quotes,attributes"]
 ----
-# {package-install-project} rhel-system-roles
+# {project-package-install} rhel-system-roles
```
Comment (ekohl) [inline guides/common/modules/proc_adding-rhel-system-roles.adoc:36]: I'm reading the guide and everything suggests this should be on the server (so `project`) is correct, but that means we only need to care about RHEL 8, since we no longer run on RHEL 7.

Could you submit that as a separate PR so we can easily cherry pick the fix to the appropriate branches?

File: guides/common/attributes-base.adoc [hunk 3]
```diff
@@ -77,8 +77,8 @@
 :oVirtEngine: {oVirt} Engine
 :oVirtShort: {oVirt}
 :package-clean: dnf clean
-:package-install-project: dnf install
-:package-install: dnf install
+:project-package-install: dnf install
+:client-package-install: dnf install
```
Comment (ekohl) [inline guides/common/attributes-base.adoc:81]: While I think this is a better name than we had (since it's qualified at least a bit), I do wonder about what the client is. I worry that we won't have "a" client, but will always need to list a few ways for the various clients.

File: guides/common/modules/proc_cloning_satellite_server.adoc [hunk 4]
```diff
@@ -177,7 +177,7 @@ You can either mount the shared storage or copy the backup files to the `/backup
 +
 [options="nowrap" subs="attributes"]
 ----
-# {package-install} satellite-clone
+# {client-package-install} satellite-clone
```
Comment (ekohl) [inline guides/common/modules/proc_cloning_satellite_server.adoc:180]: This bit is going to be confusing: cloning happens on a server, but at this point you don't have `{foreman-maintain}` installed. I'll leave a longer explanation in the whole review.

File: guides/common/modules/proc_configuring-an-external-dhcp-server.adoc [hunk 5]
```diff
@@ -18,7 +18,7 @@ include::snip_firewalld.adoc[]
 +
 [options="nowrap" subs="+quotes,attributes"]
 ----
-# {package-install} dhcp bind
+# {client-package-install} dhcp bind
```
Comment (ekohl) [inline guides/common/modules/proc_configuring-an-external-dhcp-server.adoc:21]: This isn't a client, but it's a server that's not using foreman-maintain.

File: guides/common/modules/proc_configuring-an-external-dhcp-server.adoc [hunk 6]
```diff
@@ -120,7 +120,7 @@ _990_
 +
 [options="nowrap" subs="+quotes,attributes"]
 ----
-# {package-install} {nfs-server-package}
+# {client-package-install} {nfs-server-package}
```
Comment (ekohl) [inline guides/common/modules/proc_configuring-an-external-dhcp-server.adoc:123]: Same here

File: guides/common/modules/proc_configuring-ipxe-to-reduce-provisioning-times.adoc [hunk 7]
```diff
@@ -59,7 +59,7 @@ ifdef::foreman-el,katello,satellite,orcharhino[]
 +
 [options="nowrap" subs="+quotes,attributes"]
 ----
-# {package-install} ipxe-bootimgs
+# {client-package-install} ipxe-bootimgs
```
Comment (ekohl) [inline guides/common/modules/proc_configuring-ipxe-to-reduce-provisioning-times.adoc:62]: This is on a server, so technically this should be 
```suggestion
# {project-package-install} ipxe-bootimgs
```

Because they are the same for all non-satellite, it doesn't matter but it happens on the server.

File: guides/common/modules/proc_configuring-ipxe-to-reduce-provisioning-times.adoc [hunk 8]
```diff
@@ -72,7 +72,7 @@ ifdef::foreman-deb[]
 +
 [options="nowrap" subs="+quotes,attributes"]
 ----
-# {package-install} ipxe
+# {client-package-install} ipxe
```
Comment (ekohl) [inline guides/common/modules/proc_configuring-ipxe-to-reduce-provisioning-times.adoc:75]: Same here
```suggestion
# {project-package-install} ipxe
```

**Review by ekohl — COMMENTED** [review]
Today you use `{package-install}` while doing basic installations and `{package-install-project}` once `{foreman-maintain}` is installed. The explanation can be confusing, but on Satellite we use a foreman-maintain feature where we lock all package versions, to prevent accidental upgrades. This package locking feature is not used outside of Satellite, so `{package-install}` and `{package-install-project}` are the same.

You also use plain `{package-install}` to install foreman-maintain, indirectly via the `satellite` or `capsule` meta package.

I didn't go through all of the files, but I wanted to clarify that now.

**Comment by asteflova** [comment]
@ekohl This is amazing! Thanks a lot for the review and the explanations.

@maximiliankolb After trying to make this work and reading Ewoud's comments, I'm starting to think it's not a good idea to introduce two different attributes: client-package-install and project-package-install. A lot of times, I just won't be able to tell which one to use, and I don't think it's just because I'm new to the product. Maybe the other writers on the team would be able to tell, but anyone who's new or just not that experienced would have trouble doing so. Which would be a burden on the more experienced writers and SMEs to correct the junior members. Not to mention SMEs would first need to be explained what the difference is and that they need to check the attrs. What do you think?

**Comment by maximiliankolb** [comment]
I think esp. because it is hard, we should do it. I believe that renaming the attribute is the first logical step (a simple regex as per your commits) which results in no diff in the HTML output; see first comment.

This means that all of Ewouds comments are important and must be applied because they hint towards mistakes in the docs, esp. for foreman-deb.

I didn't have time this week yet, but your work on this is definitely appreciated. I will try to have a closer look at it next week maybe.

**Comment by asteflova** [comment]
I'll apply Ewoud's suggestions next week and take more time to think on them. I will do my best to get better understanding of the difference, maybe things will get clearer to me after that.

In the meantime, thanks everyone, I really appreciate the input.

**Comment by ekohl** [comment]
You could consider `{project-package-install-bootstrap}` and `{project-package-install}` to distinguish that one happens on the project while "bootstrapping" (making sure you can use the regular tooling) and one where you have the tooling available. Of course these names aren't perfect, so take them as a hint on the direction and not the exact naming.

**Comment by asteflova** [comment]
Tagging the other doc team members.

This is a draft pull request to implement changes to how install attributes are used in the docs set. The goal is to be able to differentiate between commands run on the server ({project-package-install}) and the client ({client-package-install}).

To sumarize the current status of the PR:
* Ewoud posted helpful feedback, to be implemented
* I [expressed concern](https://github.com/theforeman/foreman-documentation/pull/2259#issuecomment-1614140983) about the practical part of being able to maintain this system
* Max believes this is [very much worth doing](https://github.com/theforeman/foreman-documentation/pull/2259#issuecomment-1614213409) to ensure the docs are accurate

@adamlazik1 @AkshayGadhaveRH @bangelic @Lennonka @mjivraja  If you have any ideas or comments, please share them.

File: guides/common/attributes-base.adoc (hunk 3 above)
Comment (Lennonka) [reply guides/common/attributes-base.adoc:81]: Client is supposed to be a host.
And I agree that for upstream we usually need to have a list of more than a single OS, so I'm not sure how this will help.

File: guides/common/modules/proc_benchmarking-raw-db-performance.adoc [hunk 9]
```diff
@@ -6,7 +6,7 @@ To get a list of the top table sizes in disk space for both Candlepin, Foreman,
 endif::[]
 
 PGbench utility (note you may need to resize PostgreSQL data directory `{postgresql-lib-dir}` directory to 100 GiB or what does benchmark take to run) might be used to measure PostgreSQL performance on your system.
-Use `{package-install} postgresql-contrib` to install it.
+Use `{client-package-install} postgresql-contrib` to install it.
```
Comment (Lennonka) [inline guides/common/modules/proc_benchmarking-raw-db-performance.adoc:9]: @ekohl This also looks like it should be `project-`. Can you confirm?

File: guides/common/attributes-base.adoc (hunk 3 above)
Comment (Lennonka) [inline guides/common/attributes-base.adoc:81]:
```suggestion
:client-package-install-el: dnf install
:client-package-install-deb: apt-get install
```
Perhaps this would be more helpful?

File: guides/common/modules/proc_benchmarking-raw-db-performance.adoc (hunk 9 above)
Comment (ekohl) [reply guides/common/modules/proc_benchmarking-raw-db-performance.adoc:9]: That's correct.

Now that I'm reading this I'm looking at what this chapter is supposed to achieve. It's quite. Katello-centric in assuming that Candlepin and Pulp are present.

Overall I wonder what the point of the linked sizes script is. https://stackoverflow.com/questions/18907047/postgres-db-size-command says you can use `\l+` the psql utility and that does work for me:
```console
$ sudo -u postgres psql
psql (12.12)
Type "help" for help.

postgres=# \l+
                                                                    List of databases
   Name    |  Owner   | Encoding |   Collate   |    Ctype    |   Access privileges   |  Size   | Tablespace |                Description                 
-----------+----------+----------+-------------+-------------+-----------------------+---------+------------+--------------------------------------------
 foreman   | foreman  | UTF8     | en_US.utf8  | en_US.utf8  | foreman=CTc/foreman  +| 38 MB   | pg_default | 
           |          |          |             |             | =T/foreman            |         |            | 
 postgres  | postgres | UTF8     | nl_NL.UTF-8 | nl_NL.UTF-8 |                       | 7337 kB | pg_default | default administrative connection database
 template0 | postgres | UTF8     | nl_NL.UTF-8 | nl_NL.UTF-8 | =c/postgres          +| 8089 kB | pg_default | unmodifiable empty database
           |          |          |             |             | postgres=CTc/postgres |         |            | 
 template1 | postgres | UTF8     | nl_NL.UTF-8 | nl_NL.UTF-8 | postgres=CTc/postgres+| 7177 kB | pg_default | default template for new databases
           |          |          |             |             | =c/postgres           |         |            | 
(4 rows)
```

Perhaps @evgeni knows better if there's a difference.

While on the topic: since Pulp 3.22 (included in Katello 4.9) we always require postgresql-contrib anyway, but it's still needed on vanilla Foreman (without Katello).

Then about the specific PGbench part: I'm not sure what's expected of the reader.

In conclusion: what's the point of this chapter?

Comment (evgeni) [reply guides/common/modules/proc_benchmarking-raw-db-performance.adoc:9]: Well, the linked script shows table sizes, not database sizes, so the output is quite different.
Whether this amount of detail is required at that point or what the reader should do with it, I do not know.

(I also prefer not to link to random GitHub repos in docs, but that's another totally different topic).

File: guides/common/attributes-base.adoc (hunk 3 above)
Comment (mjivraja) [reply guides/common/attributes-base.adoc:81]: Can we set the value as `{package-manager} install` and have an applicable package manager attribute in `attribute-{build}`?

Comment (Lennonka) [reply guides/common/attributes-base.adoc:81]: @maximiliankolb @ekohl Opinions?

File: guides/common/modules/proc_benchmarking-raw-db-performance.adoc (hunk 9 above)
Comment (ekohl) [reply guides/common/modules/proc_benchmarking-raw-db-performance.adoc:9]: > Well, the linked script shows table sizes, not database sizes (there is `\dt+` for table sizes, but that's still less detail than the script), so the output is quite different.

Ah, good point.

> Whether this amount of detail is required at that point or what the reader should do with it, I do not know.

I also wonder what the point of this chapter is. Yes, you know table sizes. But then what do you do with it?

> (I also prefer not to link to random GitHub repos in docs, but that's another totally different topic).

Also what triggered me and why I started to look at native PostgreSQL capabilities. So you can inline it.

File: guides/common/attributes-base.adoc (hunk 3 above)
Comment (Lennonka) [reply guides/common/attributes-base.adoc:81]: @maximiliankolb Yes, please, bring it to the meeting. I don't understand very well what you have in mind.

**Comment by maximiliankolb** [comment]
I will re-use this _some day_ to differentiate between "OS of Foreman" and "OS of managed hosts".

**Comment by maximiliankolb** [comment]
Closing this draft PR in favor of https://github.com/theforeman/foreman-documentation/pull/2625 cc @asteflova
